src/sundials/sundials_context.c

```diff
@@ -44,88 +62,105 @@ int SUNContext_Create(SUNComm comm, SUNContext* sunctx)
     comm = SUN_COMM_NULL;
   }
 
-#if defined(SUNDIALS_BUILD_WITH_PROFILING) && !defined(SUNDIALS_CALIPER_ENABLED)
-  if (SUNProfiler_Create(comm, "SUNContext Default", &profiler)) return (-1);
-#endif
-
-#ifdef SUNDIALS_ADIAK_ENABLED 
-  adiak_init(&comm);
-  sunAdiakCollectMetadata();
-#endif
-
 #if SUNDIALS_LOGGING_LEVEL > 0 
 #if SUNDIALS_MPI_ENABLED
-  if (SUNLogger_CreateFromEnv(comm, &logger))
-#else
-  if (SUNLogger_CreateFromEnv(SUN_COMM_NULL, &logger))
-#endif
-  {
-    return (-1);
+  if (SUNLogger_CreateFromEnv(comm, &logger)) {
+    return SUN_ERR_LOGGER_CORRUPT;
   }
 #else
-#if SUNDIALS_MPI_ENABLED
-  if (SUNLogger_Create(comm, 0, &logger))
-#else
-  if (SUNLogger_Create(SUN_COMM_NULL, 0, &logger))
+  if (SUNLogger_CreateFromEnv(SUN_COMM_NULL, &logger)) {
+    return SUN_ERR_LOGGER_CORRUPT;
+  }
 #endif
-  {
-    return (-1);
+#else
+  if (SUNLogger_Create(NULL, 0, &logger)) {
+    return SUN_ERR_LOGGER_CORRUPT;
   }
-  SUNLogger_SetErrorFilename(logger, "");
-  SUNLogger_SetWarningFilename(logger, "");
-  SUNLogger_SetInfoFilename(logger, "");
-  SUNLogger_SetDebugFilename(logger, "");
+  SUNCheckCall(SUNLogger_SetErrorFilename(logger, ""));
+  SUNCheckCall(SUNLogger_SetWarningFilename(logger, ""));
+  SUNCheckCall(SUNLogger_SetInfoFilename(logger, ""));
+  SUNCheckCall(SUNLogger_SetDebugFilename(logger, ""));
 #endif
 
-  *sunctx = NULL;
-  *sunctx = (SUNContext)malloc(sizeof(struct _SUNContext));
-
-  if (*sunctx == NULL)
-  {
 #if defined(SUNDIALS_BUILD_WITH_PROFILING) && !defined(SUNDIALS_CALIPER_ENABLED)
-    SUNProfiler_Free(&profiler);
+  SUNCheckCall(SUNProfiler_Create(comm, "SUNContext Default", &profiler));
 #endif
-    SUNLogger_Destroy(&logger);
-    return (-1);
-  }
 
-  (*sunctx)->logger       = logger;
-  (*sunctx)->own_logger   = logger != NULL;
-  (*sunctx)->profiler     = profiler;
-  (*sunctx)->own_profiler = profiler != NULL;
+  sunctx->logger        = logger;
+  sunctx->own_logger    = logger != NULL;
+  sunctx->profiler      = profiler;
+  sunctx->own_profiler  = profiler != NULL;
+  sunctx->last_err      = 0;
+  sunctx->err_handler   = SUNErrHandler_Create(SUNLogErrHandlerFn, NULL);
+  sunctx->comm          = comm;
+
+  *sunctx_out = sunctx;
 
-  return (0);
+  return SUN_SUCCESS;
 }
 
-int SUNContext_GetProfiler(SUNContext sunctx, SUNProfiler* profiler)
+SUNErrCode SUNContext_GetLastError(SUNContext sunctx, SUNErrCode* last_err)
 {
-  if (sunctx == NULL)
+  *last_err = sunctx->last_err;
+  sunctx->last_err = SUN_SUCCESS;
```

Comment:
I prefer the `GetLastError` and `PeekLastError` approach for the follwing reasons:

1. This aligns with the [CUDA API](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__ERROR.html#group__CUDART__ERROR_1ge9f0596369c32aa43875c17c78a8fe41) 
2. Having `GetLastError` clear the error creates a functionality similar to returning an error code from the function. What I mean is this: when a function returns an error code, the value returned is now in the hands of the user. If they dont want to handle it right away, then they need to store it in some variable or ignore it. If `GetLastError` returns the error code and then resets it, this creates the same situation for the user (they must handle it or store it for later or ignore it). If `GetLastError` was to not reset the error, then it may encourage the user to rely on the error code still being available later which may be untrue if another error overwrites it.

Reply:
Thanks for clarifying the reasoning. This sounds good to me and we can keep these functions as they are now.